pyro/infer/mcmc/hmc.py

```diff
@@ -94,45 +149,65 @@ def setup(self, *args, **kwargs):
         self._kwargs = kwargs
         # set the trace prototype to inter-convert between trace object
         # and dict object used by the integrator
-        self._prototype_trace = poutine.trace(self.model).get_trace(*args, **kwargs)
+        trace = poutine.trace(self.model).get_trace(*args, **kwargs)
+        self._prototype_trace = trace
         # momenta distribution - currently standard normal
-        for name, node in sorted(self._prototype_trace.iter_stochastic_nodes(),
-                                 key=lambda x: x[0]):
-            r_mu = torch.zeros_like(node['value'])
-            r_sigma = torch.ones_like(node['value'])
+        for name, node in sorted(trace.iter_stochastic_nodes(), key=lambda x: x[0]):
+            r_mu = torch.zeros_like(node["value"])
+            r_sigma = torch.ones_like(node["value"])
             self._r_dist[name] = dist.Normal(mu=r_mu, sigma=r_sigma)
-            if node['fn'].support is not constraints.real and self._automatic_transform_enabled:
-                self.transforms[name] = biject_to(node['fn'].support).inv
-        self._validate_trace(self._prototype_trace)
+        if node["fn"].support is not constraints.real and self._automatic_transform_enabled:
```

Comment:
This should be inside the `for` block as we want to populate transforms for each of the nodes, if applicable.

As an aside, it is always nice to separate out the refactoring related changes from the issue-specific changes, and address them in separate PRs. That makes it easier to review both of them. No need to worry about this one though, as I can see that it will be tough to do with all the merges. :)